src/mlpack/methods/ann/layer/max_pooling.hpp

```diff
@@ -171,18 +171,34 @@ class MaxPoolingType : public Layer<MatType>
         for (size_t i = 0, rowidx = 0; i < output.n_rows;
             ++i, rowidx += strideWidth)
         {
+          size_t rowEnd = rowidx + kernelWidth - 1;
+          size_t colEnd = colidx + kernelHeight - 1;
+
+          if (rowEnd > input.n_rows - 1)
+          {
+            if (floor)
+              continue;
+            rowEnd = input.n_rows - 1;
+          }
+
+          if (colEnd > input.n_cols - 1)
+          {
+            if (floor)
+              continue;
+            colEnd = input.n_cols - 1;
+          }
           const std::tuple<size_t, typename MatType::elem_type> poolResult =
               pooling.PoolingWithIndex(input.slice(s).submat(
                   rowidx,
                   colidx,
-                  rowidx + kernelWidth - 1 - offset,
-                  colidx + kernelHeight - 1 - offset));
+                  rowEnd,
+                  colEnd));
 
           // Now map the returned pooling index, which corresponds to the
           // submatrix we gave, back to its position in the (linearized) input.
           const size_t poolIndex = std::get<0>(poolResult);
-          const size_t poolingCol = poolIndex / (kernelWidth - offset);
-          const size_t poolingRow = poolIndex % (kernelWidth - offset);
+          const size_t poolingCol = poolIndex / (kernelWidth);
+          const size_t poolingRow = poolIndex % (kernelWidth);
```

Comment:
Do you mind providing a quick explanation of what was wrong before and how this fixes it?  (Just to make sure that I am understanding correctly as I review.)

Reply:
Here, if you set kernel size of 1 and then offset is set 1 too. Then we will have a divide error. The offset variable was used long ago, and its use was removed after you started working on #2777. But it somehow made it into the code.

Reply:
That's true, but could that situation ever happen?  My memory of `offset` is that it is meant to indicate whether a pooling operation should start at a row/col that's not 0.  How is this handled now?  Is that not possible in the new code?  (It's not a problem if it isn't, but I do want to know how we handle the case where the input size is not divisible by the kernel.)

Reply:
I don't think so that offset does this. If we go by the previous code, it will make no sense to modify the kernel size of all layers. Now, you can't offset the start in this code. For handling the case where the input size is not divisible by the kernel, we are using floor and ceil to determine the size of the kernel at the last block. You can see there are some if-else code for handling the edge cases as well.